### DISCUSSION TITLE
x-modal with index iterator in namepace inside template

### DISCUSSION THREAD

**Discussion by d3cline** [alpinejs/alpine#1096] [1. Help]
Say I have many input boxes like this, 

```
  <template x-if="t == 'forward'" >
      <div>
        <label for="email" class="block text-sm font-medium text-gray-700">Forward Address</label>
        <div class="mt-1">
          <input type="email" x-model="'add_forward_'+index" :id="'add_forward_'+index" class="my-4 shadow-sm focus:ring-indigo-500 focus:border-indigo-500 block w-full sm:text-sm border-gray-300 rounded-md" placeholder="you@example.com" aria-describedby="email-description" >
        </div>
      </div>
  </template>
```

How can I use x-model here correctly? The above half works. I can see the iterator in the initial state of the input. Upon changing the input the following error renders in the console, 
![image](https://user-images.githubusercontent.com/77483/108035107-b9361f00-6ff3-11eb-9bc6-9d59045bf09a.png)

```
Uncaught SyntaxError: invalid assignment left-hand side
```

This appears to be something in the library itself. I am cheating upstream to see if this will work at all using jinja template, 

```
    {% for i in '0123456789'%}add_forward_{{i}}: null,{%endfor%}
```

But this is a bad solution and doesn't work anyway. I tried assigning this as an array as opposed to null but it also didn't work, it would bind all inputs to a single model.

What am I doing wrong? Ultimately I am trying to do form validation so maybe there is an easier way, but I should be able to perform any task upstream if the iterator is correctly passed. 

P.S. any way I could be reading it better with `$ref`, see how I am doing that with ID, I can read it later using `document.` but could this be better as well?

**Comment by SimoTod** [comment]
x-model needs a variable not a string. This is because it gets translated to
`variable = ...`. Your code would become
`'add_forward_'+index = ...` which is not a valid assignment.
You should be able to use an array and have `x-model="add_forward[index]"`.

**Reply by d3cline** [reply]
This seems to have progressed things, I am now rendering `undefined` 
![image](https://user-images.githubusercontent.com/77483/108074793-d59f7f00-7026-11eb-9826-b49a4b29baab.png)

This happens if I change the variable in `x-data` function into an array. So it appears to be assigning the var someplace, but still seems like its failing upstream in the module.

**Reply by d3cline** [reply]
As a guess I wanted to try to assign strings to that array and this happened, 
![image](https://user-images.githubusercontent.com/77483/108075504-ae957d00-7027-11eb-8632-479ee5e71f29.png)
So I suppose its 'working' although it seems like its still not quite right that I would have to populate the array with a bunch of empty strings, 

```
    add_forward: ['','',''],
```

P.S. I am more of python dev and maybe I am not having the correct expectation here. Is this the right track?

**Reply by KevinBatdorf** [reply]
> So I suppose its 'working' although it seems like its still not quite right that I would have to populate the array with a bunch of empty strings,

Sounds like a good PR to make if you're up for it

**Reply by d3cline** [reply]
I have been thinking of ways to make this automatically populate the array. If I figure it out I will.

**Reply by SimoTod** [reply]
It's tricky, if the value is undefined, one would expect the UI to show an empty string. On the other end, being a two ways sync, you would risk to override the original value (undefined to '') which should be fine but I don't know if it would break things for anyone relying on this weird behaviour. Probably one for @calebporzio to decide if he would accept a PR. 

@d3cline To init the array, you can use `Array(n).fill('')`
